### PR TITLE
`Variables` module

### PR DESCRIPTION
**Pull request by jakekaplan**
Adds `Variables` to be used in and outside of flows.

You can access variables in 3 ways:

```python
from prefect import variables

@flow
async def my_flow():
    variables.get("my_var")
    variables["my_var"]
    await variables.get("my_var")
```
   
closes: https://github.com/PrefectHQ/prefect/issues/9108

### CODE REVIEW

File: tests/test_variables.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+):
+    model = await create_variable(
+        session,
+        VariableCreate(name="my_variable", value="my-value", tags=["123", "456"]),
+    )
+    await session.commit()
+
+    return model
+
+
```
Comment (jakekaplan) [inline tests/test_variables.py:21]: I am not totally sure what else would make sense to test here. Suggestions welcome!

File: src/prefect/variables.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from prefect.client.orchestration import get_client
+
+
+def get(name: str, default: str = None) -> Optional[str]:
+    """Get a variable by name from a sync context"""
+    variable = get_variable_by_name(name)
+    return variable.value if variable else default
+
+
+async def aget(name: str, default: str = None) -> Optional[str]:
```
Comment (jakekaplan) [inline src/prefect/variables.py:15]: @madkinsz is this what you had in mind when you said offering something that users could await?

Comment (madkinsz) [reply src/prefect/variables.py:25]: Technically it should probably use `sync_compatible` — we don't really publish two interfaces as a general rule.

File: src/prefect/variables.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        async def my_flow():
+            var = await variables.aget("my_var")
+    ```
+    """
+    variable = await _get_variable_by_name(name)
+    return variable.value if variable else default
+
+
+async def _get_variable_by_name(name):
+    async with get_client() as client:
```
Comment (madkinsz) [inline src/prefect/variables.py:42]: Should probably use `inject_client` so it receives optimizations to client retrieval

File: src/prefect/variables.py (hunk 2 above)
Comment (jakekaplan) [reply src/prefect/variables.py:25]: duh 🤦‍♂️, good call

**Comment by billpalombi** [comment]
Access patterns lgtm!

**Comment by jakekaplan** [comment]
Does this need to get added to `__init__` under the user facing api?

File: src/prefect/variables.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+def get_variable_by_name(name: str):
+    variable = from_sync.call_soon_in_loop_thread(
+        create_call(_get_variable_by_name, name)
+    ).result()
+
+    return variable
+
+
+def __getitem__(name: str) -> Optional[str]:
+    # This is a stub for VariablesModule.__getitem__
```
Comment (jakekaplan) [inline src/prefect/variables.py:53]: this makes IDE's happy but not sure if this is kosher

File: src/prefect/variables.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            @flow
+            def my_flow():
+                var = variables["my_var"]
+        ```
+        """
+        variable = get_variable_by_name(name)
+        return variable.value if variable else None
+
+
+sys.modules[__name__].__class__ = VariablesModule
```
Comment (madkinsz) [inline src/prefect/variables.py:74]: This is a little terrifying. Why not just implement this as a `__getitem__` without injecting it into the modules and defining a class?

Comment (madkinsz) [reply src/prefect/variables.py:74]: Do modules not allow `__getitem__` just `__getattr__` otherwise?

File: src/prefect/variables.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+@inject_client
+async def _get_variable_by_name(
+    name: str,
+    client: PrefectClient,
+):
+    variable = await client.read_variable_by_name(name)
+    return variable
+
+
+def get_variable_by_name(name: str):
```
Comment (madkinsz) [inline src/prefect/variables.py:44]: Should this be private too for now?

File: src/prefect/variables.py (hunk 5 above)
Comment (madkinsz) [reply src/prefect/variables.py:74]: I see... https://stackoverflow.com/questions/29064125/add-getitem-to-module

This is quite the hack. Do we care enough about the `["name"]` interface to do this?

File: src/prefect/variables.py (hunk 4 above)
Comment (madkinsz) [reply src/prefect/variables.py:53]: Cool that this works for type checkers :D nice

File: src/prefect/variables.py (hunk 5 above)
Comment (madkinsz) [reply src/prefect/variables.py:74]: For the record, this shouldn't hold up the PR. I think we should probably expose this kind of interface in a follow-up so we can figure out if it's worth the possible complications but I'm okay with whatever

**Comment by madkinsz** [comment]
> Does this need to get added to `__init__` under the user facing api?

Might as well add it there. We don't have a clear public/private split as things are setup right now.

File: src/prefect/variables.py (hunk 5 above)
Comment (jakekaplan) [reply src/prefect/variables.py:74]: @billpalombi are you okay if we remove accessing variables like `variables["my_var"]` for now? I don't think this it's too bad but @madkinsz is right it's probably the safer route to add later vs now and then have to remove it

Comment (billpalombi) [reply src/prefect/variables.py:74]: Yup, I'm totally fine with that.

**Pull request merged**